src/FrameBase.re

```diff
@@ -60,30 +60,65 @@ external waitForXPath :
   Js.Promise.t(ElementHandle.t) =
   "";
 
-/* TODO: Add support [, ...args] */
+type jsonOrHandle =
+  | JSON(serializable)
+  | Handle(JSHandle.t);
+
+/**
+ * selectOneEval(selector, fn, args, page)
+ * Runs document.querySelector in the page and passes it as the first
+ * argument to `fn`. Additional arguments to the `fn` can be provides in
+ * `args`. If there's no element matching selector, the method throws an
+ * error. If the function returns a Promise, it waits for the promise to
+ * resolve and returns its value.
+ */
 [@bs.send.pipe : t]
-external selectOneEval : (string, unit => unit) => Js.Promise.t('a) = "$eval";
-
-/* TODO: Add support [, ...args] */
+[@bs.splice]
+external selectOneEval :
+  (string, 'a => 'b, array(jsonOrHandle)) => Js.Promise.t(serializable) =
+  "$eval";
+
+/**
+ * selectAllEval(selector, fn)
+ * Runs document.querySelectorAll in the page and passes it as the first
+ * argument to `fn`. If `fn` returns a Promise, then it waits for
+ * the promise to resolve and returns its value.
+ */
 [@bs.send.pipe : t]
-external selectAllEval : (string, unit => unit) => Js.Promise.t('a) = "$$eval";
+external selectAllEval :
+  (string, Dom.element => 'a) => Js.Promise.t(serializable) =
+  "$$eval";
+
+[@bs.send.pipe : t] [@bs.splice]
+external selectAllEvalWithArgs :
+  (string, Dom.element => 'a, array(jsonOrHandle)) =>
+  Js.Promise.t(serializable) =
+  "$$eval";
 
 [@bs.send.pipe : t]
 external addScriptTag : tagOptions => Js.Promise.t(ElementHandle.t) = "";
 
 [@bs.send.pipe : t]
 external addStyleTag : tagOptions => Js.Promise.t(ElementHandle.t) = "";
 
-/* TODO: Currently only ever work for functions taking no arguments,
-   and the second parameter array can only ever be empty */
 [@bs.send.pipe : t] [@bs.splice]
 external evaluate :
-  (unit => Js.Promise.t(Js.Json.t), array({..})) => Js.Promise.t(Js.Json.t) =
+  (Eval.Fn.t, array(Eval.Arg.t)) => Js.Promise.t(serializable) =
   "";
 
-/* TODO: Currently only ever work for functions taking no arguments,
-   and the second parameter array can only ever be empty */
+/**
+ * Evaluates a JavaScript expression in page and returns result in a promise.
+ */
+[@bs.send.pipe : t]
+external evaluateString : string => Js.Promise.t('a) = "evaluate";
+
 [@bs.send.pipe : t] [@bs.splice]
 external evaluateHandle :
-  (unit => Js.Promise.t(JSHandle.t), array({..})) => Js.Promise.t(JSHandle.t) =
+  ('a => Js.Promise.t(JSHandle.t), array(Eval.Arg.t)) =>
```

Comment:
The first arg to `evaluateHandle` should be a function of any number of arguments (equal to the length of the array in the second arg to evaluateHandle, though I don't know if we can enforce that). Again, need to investigate whether we want/can require the user to curry the eval function with args so that we just take a function of type `unit => Js.Promise.t(JSHandle.t)`.